swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,927 @@
+    UpdateState(Arc<TState>),
+    NewOutbound(TOutboundOpenInfo),
+}
+
+pub struct FromFnProto<TInbound, TOutbound, TOutboundOpenInfo, TState> {
+    protocol: &'static str,
+
+    on_new_inbound: Option<
+        Box<
+            dyn Fn(
+                    NegotiatedSubstream,
+                    PeerId,
+                    ConnectedPoint,
+                    Arc<TState>,
+                ) -> BoxStream<'static, TInbound>
+                + Send,
+        >,
+    >,
+    on_new_outbound: Box<
+        dyn Fn(
+                NegotiatedSubstream,
+                PeerId,
+                ConnectedPoint,
+                Arc<TState>,
+                TOutboundOpenInfo,
+            ) -> BoxStream<'static, TOutbound>
+            + Send,
+    >,
+
+    inbound_streams_limit: usize,
+    pending_outbound_streams_limit: usize,
+
+    state: Arc<TState>,
+}
+
+impl<TInbound, TOutbound, TOutboundOpenInfo, TState> IntoConnectionHandler
+    for FromFnProto<TInbound, TOutbound, TOutboundOpenInfo, TState>
+where
+    TInbound: fmt::Debug + Send + 'static,
+    TOutbound: fmt::Debug + Send + 'static,
+    TOutboundOpenInfo: fmt::Debug + Send + 'static,
+    TState: fmt::Debug + Send + Sync + 'static,
+{
+    type Handler = FromFn<TInbound, TOutbound, TOutboundOpenInfo, TState>;
+
+    fn into_handler(
+        self,
+        remote_peer_id: &PeerId,
+        connected_point: &ConnectedPoint,
+    ) -> Self::Handler {
+        FromFn {
+            protocol: self.protocol,
+            remote_peer_id: *remote_peer_id,
+            connected_point: connected_point.clone(),
+            inbound_streams: SelectAll::default(),
+            outbound_streams: SelectAll::default(),
+            on_new_inbound: self.on_new_inbound,
+            on_new_outbound: self.on_new_outbound,
+            inbound_streams_limit: self.inbound_streams_limit,
+            pending_outbound_streams: VecDeque::default(),
+            pending_outbound_streams_limit: self.pending_outbound_streams_limit,
+            failed_open: VecDeque::default(),
+            state: self.state,
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ConnectionHandler>::InboundProtocol {
+        if self.on_new_inbound.is_some() {
+            EitherUpgrade::B(ReadyUpgrade::new(self.protocol))
+        } else {
+            EitherUpgrade::A(DeniedUpgrade)
+        }
+    }
+}
+
+pub struct FromFn<TInbound, TOutbound, TOutboundInfo, TState> {
+    protocol: &'static str,
+    remote_peer_id: PeerId,
+    connected_point: ConnectedPoint,
+
+    inbound_streams: SelectAll<BoxStream<'static, TInbound>>,
+    outbound_streams: SelectAll<BoxStream<'static, TOutbound>>,
+
+    on_new_inbound: Option<
+        Box<
+            dyn Fn(
+                    NegotiatedSubstream,
+                    PeerId,
+                    ConnectedPoint,
+                    Arc<TState>,
+                ) -> BoxStream<'static, TInbound>
```

Comment:
I’d prefer type aliases for these function types

Reply:
What do you mean exactly? As a user, you shouldn't need to come in contact this this `BoxStream` per se. Where would you like to see a type alias?

Reply:
I only mean the readability of this source module by itself.